exporter/collector/logs.go

```diff
@@ -43,20 +62,149 @@ func (l *LogsExporter) Shutdown(ctx context.Context) error {
+type httpRequestLog struct {
+	RequestMethod                  string `json:"requestMethod"`
+	RequestURL                     string `json:"requestUrl"`
+	RequestSize                    int64  `json:"requestSize,string"`
+	Status                         int    `json:"status,string"`
+	ResponseSize                   int64  `json:"responseSize,string"`
+	UserAgent                      string `json:"userAgent"`
+	RemoteIP                       string `json:"remoteIp"`
+	ServerIP                       string `json:"serverIp"`
+	Referer                        string `json:"referer"`
+	Latency                        string `json:"latency"`
+	CacheLookup                    bool   `json:"cacheLookup"`
+	CacheHit                       bool   `json:"cacheHit"`
+	CacheValidatedWithOriginServer bool   `json:"cacheValidatedWithOriginServer"`
+	CacheFillBytes                 int64  `json:"cacheFillBytes,string"`
+	Protocol                       string `json:"protocol"`
+}
+
+func parseHttpRequest(message []byte) (*logging.HTTPRequest, []byte, error) {
+	parsedLog, strippedMessage, err := extractHttpRequestFromLog(message)
+	if err != nil {
+		return nil, message, err
+	}
+
+	req, err := http.NewRequest(parsedLog.RequestMethod, parsedLog.RequestURL, nil)
+	if err != nil {
+		return nil, message, err
+	}
+	req.Header.Set("Referer", parsedLog.Referer)
+	req.Header.Set("User-Agent", parsedLog.UserAgent)
+
+	httpRequest := &logging.HTTPRequest{
+		Request:                        req,
+		RequestSize:                    parsedLog.RequestSize,
+		Status:                         parsedLog.Status,
+		ResponseSize:                   parsedLog.ResponseSize,
+		LocalIP:                        parsedLog.ServerIP,
+		RemoteIP:                       parsedLog.RemoteIP,
+		CacheHit:                       parsedLog.CacheHit,
+		CacheValidatedWithOriginServer: parsedLog.CacheValidatedWithOriginServer,
+		CacheFillBytes:                 parsedLog.CacheFillBytes,
+		CacheLookup:                    parsedLog.CacheLookup,
+	}
+	if parsedLog.Latency != "" {
+		latency, err := time.ParseDuration(parsedLog.Latency)
+		if err == nil {
+			httpRequest.Latency = latency
+		}
+	}
+
+	return httpRequest, strippedMessage, nil
+}
+
+func extractHttpRequestFromLog(message []byte) (*httpRequestLog, []byte, error) {
+	httpRequestKey := "httpRequest" // TODO(@braydonk) Should this come from the config?
+
+	unmarshalledMessage := make(map[string]interface{})
+	if err := json.Unmarshal(message, &unmarshalledMessage); err != nil {
+		return nil, message, err
+	}
+
+	httpRequestMap, ok := unmarshalledMessage[httpRequestKey]
+	if !ok {
+		return nil, message, fmt.Errorf("message has no key %s", httpRequestKey)
+	}
+	httpRequestBytes, err := json.Marshal(httpRequestMap)
+	if err != nil {
+		return nil, message, err
+	}
+	var httpRequest *httpRequestLog
+	if err := json.Unmarshal(httpRequestBytes, &httpRequest); err != nil {
```

Comment:
instead of marshalling and unmarshalling, can you just cast the map directly like `httpRequest := httpRequestMap.(*httpRequestLog)`?

Reply:
Obsolete with [7903140](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/344/commits/79031407118a05e991071d25953f06de42894659)